### ISSUE TITLE
Detect user's color theme preference

### ISSUE THREAD

**Issue by Fausto95** [excalidraw/excalidraw#2025]
Since [Dark mode](https://github.com/excalidraw/excalidraw/pull/2006) has landed, it would be nice if we could use system's color theme as default `prefers-color-scheme`

**Comment by xixixao** [comment]
I considered this (including providing a three-way toggle), and I've landed on this being the simplest, most user friendly solution. Because this doesn't just change the UI (that would be ok to default), but also the canvas, I don't think we should default to dark mode based on prefers-color-scheme.

**Comment by dwelle** [comment]
That's what `prefers-color-scheme` is for --- doesn't really matter what our Dark mode includes in the package.

That said, our current implem doesn't indeed allow to differentiate between user & browser preference so we can't default to anything without a rewrite.

**Comment by tomayac** [comment]
There is a custom element (full disclosure: I built it) called [`<dark-mode-toggle>`](https://github.com/GoogleChromeLabs/dark-mode-toggle) that initially respects the user's system preference (via `prefers-color-scheme`), but that then allows them to override it permanently, or temporarily for the session. It supports the `Appearance_dark` class approach ([docs](https://github.com/GoogleChromeLabs/dark-mode-toggle#-using-a-css-class-that-you-toggle)). You can see it used on [v8.dev](https://v8.dev/). I'm happy to PR it into Excalidraw (but even happier for someone else to try it).

**Comment by tomayac** [comment]
(I just fixed a small issue with the [brilliant] dark mode implementation, and the lack of `prefers-color-scheme` support was the first thing that I noted when I toggled dark mode in the simulator.)

**Comment by lipis** [comment]
Seems minimal.. I wouldn't mind using it @tomayac.. and I'm happy if you PR it into Excalidraw.. /cc @dwelle

**Comment by dwelle** [comment]
@tomayac without digging deep, does your compo support all the above behavior without the need for the "permanent" checkbox? In other words, the behavior would be as such: on first visit (or any subsequent visits until user interacts with the button), use `prefers-color-scheme`, and once user toggles the button, use the localStorage value from then on?

**Comment by xixixao** [comment]
I'll just repeat for posterity and then leave this to whoever wants to work on it:

I have considered whether the dark mode should be automatic, or whether we should have a 3-way toggle (one option for preserving system preference). I have opted for the current design for 2 reasons:

1) It's simple.
2) Even people with Dark Mode on in their OS might still want a "whiteboard", not a "blackboard". Other apps do this too, for example MacOS Pages - when you create a new document, the default theme has white background, regardless of what your system dark mode setting is. And it didn't make sense to me to separate the dark mode for UI from dark mode for canvas (again point 1) ).

**Comment by dwelle** [comment]
Yes, I don't think we want 3-way toggle nor checkboxes or anything like that. If the non-traditional approach I suggested feels not good enough I'm for closing this and not introducing additional complexity to the UI.

**Comment by tomayac** [comment]
Looking at the code, it seems like the simplest solution would be to initialize the default state based on the user's current preference. https://github.com/excalidraw/excalidraw/pull/2126 does exactly this.

**Comment by robertvanhoesel** [comment]
Would an other valid suggestion be to point out the toggle in the UI whenever `prefers-colorscheme: dark` is detected? Not sure if there's already a fitting UI pattern for that.

**Comment by giuscri** [comment]
@robertvanhoesel You mean: if `prefers-colorscheme: dark` is detected, **turn on dark mode** and _point out_ (how?) the toggle?

**Comment by robertvanhoesel** [comment]
@giuscri exactly! Like a little toast or something "We switched to dark mode based on your system setting, click 🌙 to change to light mode".

Its a nice way to discover the option.

**Comment by giuscri** [comment]
@xixixao Do you think you would accept a PR that implements the behaviour @robertvanhoesel is describing?

**Comment by xixixao** [comment]
We could start with pointing out dark mode and not turning it on. Although I'm not sure what precedent this sets for NUX in general (new user experience).

**Comment by dwelle** [comment]
I don't think we want any popups or toasts, especially for new users. In this respect, Excalidraw is pretty adamant.

As for whether to automatically switch — IIRC the previous consensus was not to do this so as to present Excalidraw in the color scheme it is branded in and is the default.

What's the argument here? If we think the color scheme toggle is too hidden, we may want to put it in the top-right corner as is standard.

**Comment by giuscri** [comment]
The argument is that when you open the laptop lid in the dark, if
Excalidraw is the last thing you left open (in full screen!) it will flash
your eyes and make you hate the tool

This is most likely a user problem in all the other cases, but the amount
of white in a typical Excalidraw session is very high. To make things worse
I use Excalidraw in full screen a lot of times, especially when
collaborating with the rest of the teams.

**Comment by robertvanhoesel** [comment]
My suggestion was purely for making discovery of the feature easier to users already using darkmode, like #2126. The argument for a clean simple UI to me weighs much stronger so I agree with @dwelle on this, keep things as they are. (I do think the position is fine)

**Comment by xixixao** [comment]
@giuscri can you explain a bit more your use case? Do you have time-based dark mode toggle set in your OS? Which OS are you using? Why not always use dark mode "on"?

**Comment by giuscri** [comment]
@xixixao Sure: I'm using macOS and a time-based dark mode toggle setup. I like using _light_ mode during the day as I feel I'm more productive using it. But I like using dark mode when productivity is not a priority instead (during evenings and night)

**Comment by xixixao** [comment]
@giuscri makes sense. I'm not sure whether we can prevent a flash, given you're just opening the laptop after the dark mode switched.

Nevertheless, it would make sense to me follow the preference IF it has changed. So,

1. First time users should always see light mode
2. If you previously had light OS preference and light mode on, but now you have dark OS preference, the UI will switch to dark mode. (same the other way around)

If we get a PR we can test out how it feels.

**Comment by giuscri** [comment]
@xixixao What distinguishes a first-time user from a regular user? A regular user is one that closed Excalidraw at least once?

**Comment by dwelle** [comment]
props on @xixixao's proposal.

@giuscri yes, I'd just set some key in localStorage. If present, it means you're a returning user (provided you set it *after* you initially read it for the purpose of setting the theme).

In fact, we may wanna move `appState.theme` from `appState` to localStorage altogether (which we've discussed in a recent PR), because we'll need to do so if we want to apply the theme to the initial loading screen. But I'm not sure how it'd tie in to the fact that we want it to be controlled by the app (core). /cc @ad1992 thoughts?

**Comment by ad1992** [comment]
> In fact, we may wanna move appState.theme from appState to localStorage altogether (which we've discussed in a recent PR), because we'll need to do so if we want to apply the theme to the initial loading screen. But I'm not sure how it'd tie in to the fact that we want it to be controlled by the app (core). /cc @ad1992 thoughts?

@dwelle yes we can keep it in LS just like we keep the zen mode for example and once loaded it will be taken from appState. LS data will be used for initial load and post the change will be saved in state and in `onChange` we can keep pushing it to LS as you mentioned.

**Comment by somebody1234** [comment]
at the very least theme setting should persist, right? or does that already happen and there's something going wonky on my end?

**Comment by dwelle** [comment]
It does persist already. What's your Browser & OS? Are you using incognito mode?

**Comment by somebody1234** [comment]
no incognito mode, standard chrome 91 on linux. can confirm it doesn't persist even when just closing and reopening (via url, not via un-closing) the tab.

**Comment by accforgithubtest** [comment]
Any chances of being able to set the defaults as darkmode in the docker compose env variable ? 
As someone who always goes dark mode everything, and using docker compose to run excalidraw, it would be awesome to have this set once in compose and its always in darkmode !
